target/riscv/cpu_helper.c

```diff
@@ -584,7 +584,17 @@ static int get_physical_address(CPURISCVState *env, hwaddr *physical,
             return TRANSLATE_FAIL;
         }
 
+#if defined(TARGET_CHERI) && !defined(TARGET_RISCV32)
+        /*
+         * The top ten bits of the PTE are reserved for extensions; on CHERI,
+         * these are taken for permission bits.  (On other variants, they are
+         * presumably extensions of the ppn.)  At the moment, CHERI only uses
+         * the top five, but zero them all anyway.
+         */
+        hwaddr ppn = (pte & ~0xFFC0000000000000ULL) >> PTE_PPN_SHIFT;
```

Comment:
If the top 10 bits are reserved in the privilege spec, then this should probably be an upstream change conditional on #ifndef TARGET_RISCV32?

I'm curious how the existing PTE bits here didn't break when this was used.

Reply:
In all cases where `ppn` is used, it either checks specific bits (low bits for superpage and greater alignment) or ends up left-shifting by `PGSHIFT` which is 12, ie 2 more than what we right-shifted by, which is conveniently exactly the number of bits in the first revision.

Reply:
The top two are "accidentally" hidden: the pte is shifted down by 10 to chop off the bottom permission bits and get the ppn, but then up by 12 to make room for the page offset.

You're right; it's an upstreamable change; standby one.